src/access/examples/ExampleAccessControlDelegate.cpp

```diff
@@ -0,0 +1,1101 @@
+
+private:
+    ClusterId mCluster;
+    DeviceTypeId mDeviceType;
+};
+
+class EntryStorage
+{
+public:
+    // ACL support
+    static EntryStorage acl[kNumberOfFabrics * kEntriesPerFabric];
+
+    static EntryStorage * FindUnusedInAcl(size_t * index, const FabricIndex * fabricIndex)
+    {
+        for (auto & storage : acl)
+        {
+            if (!storage.InUse())
+            {
+                if (index != nullptr)
+                {
+                    if (fabricIndex != nullptr)
+                    {
+                        *index = 0;
+                        for (auto & storage2 : acl)
+                        {
+                            if (&storage2 == &storage)
+                            {
+                                break;
+                            }
+                            else if (storage2.mFabricIndex == storage.mFabricIndex)
+                            {
+                                (*index)++;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        *index = &storage - acl;
+                    }
+                }
+                return &storage;
+            }
+        }
+        return nullptr;
+    }
+
+    static EntryStorage * FindUsedInAcl(size_t index, const FabricIndex * fabricIndex)
+    {
+        index = FilterIndex(index, fabricIndex);
+        if (index < ArraySize(acl))
+        {
+            auto storage = acl + index;
+            if (storage->InUse())
+            {
+                return storage;
+            }
+        }
+        return nullptr;
+    }
+
+public:
+    // Pool support
+    static EntryStorage pool[kEntryStoragePoolSize];
+
+    static EntryStorage * Find(EntryStorage * candidate)
+    {
+        if (candidate && candidate->InPool())
+        {
+            return candidate;
+        }
+        for (auto & storage : pool)
+        {
+            if (!storage.InUse())
+            {
+                return &storage;
+            }
+        }
+        return nullptr;
+    }
+
+    bool InPool() const
+    {
+        constexpr auto end = pool + ArraySize(pool);
+        return pool <= this && this < end;
+    }
+
+public:
+    EntryStorage() = default;
+
+    void Init()
+    {
+        if (!mInUse)
+        {
+            Clear();
+            mInUse = true;
+        }
+    }
+
+    bool InUse() const { return mInUse; }
+
+    void Release()
+    {
+        if (InPool())
+        {
+            mInUse = false;
+        }
+    }
+
+    void Clear()
+    {
+        mInUse = false;
+        mFabricIndex = chip::kUndefinedFabricIndex;
+        mAuthMode = AuthMode::kPase;
+        mPrivilege = Privilege::kView;
+        memset(mSubjects, SubjectStorage::kEmptyByte, sizeof(mSubjects));
+#if defined(__GNUC__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wclass-memaccess"
+        memset(mTargets, TargetStorage::kEmptyByte, sizeof(mTargets));
```

Comment:
A loop to set to an empty-value will be optimized under the hood to a memset if it's legal. Suggest using the loop rather than a memset of a byte representation.

Reply:
Is that true? I'd love if it were so, but I was hoping to actually disassemble it to confirm before removing this optimization.

Reply:
I am not sure the optimization is worth the special casing and extra mental analysis. It seems like a premature optimization.

Reply:
Though the storage types are POD, I removed the memsets (can re-add later when measuring how much code they may save).